server/src/server.ts

```diff
@@ -1,172 +1,631 @@
     ) {
       supportSASGetLibList = true;
     }
+    _pyrightLanguageProvider.initialize(params, [], []);
+
     const result: InitializeResult = {
       capabilities: {
         textDocumentSync: TextDocumentSyncKind.Incremental,
         semanticTokensProvider: {
           legend,
           full: true,
         },
-        documentSymbolProvider: true,
         documentFormattingProvider: true,
         foldingRangeProvider: true,
-        hoverProvider: true,
-        completionProvider: {
-          triggerCharacters: [" "],
-          resolveProvider: true,
-        },
         documentOnTypeFormattingProvider: {
           firstTriggerCharacter: "\n",
           moreTriggerCharacter: [";"],
         },
+        documentSymbolProvider: { workDoneProgress: true },
+        workspaceSymbolProvider: { workDoneProgress: true },
+        hoverProvider: { workDoneProgress: true },
+        completionProvider: {
+          triggerCharacters: _pyrightLanguageProvider.getClientCapabilities()
+            .hasVisualStudioExtensionsCapability
+            ? [".", "[", "@", '"', "'", " "]
+            : [".", "[", '"', "'", " "],
+          resolveProvider: true,
+          workDoneProgress: true,
+          completionItem: {
+            labelDetailsSupport: true,
+          },
+        },
         signatureHelpProvider: {
-          triggerCharacters: ["(", ","],
+          triggerCharacters: ["(", ",", ")"],
+          workDoneProgress: true,
+        },
+        workspace: {
+          workspaceFolders: {
+            supported: true,
+            changeNotifications: true,
+          },
         },
       },
     };
     return result;
   });
 
+  connection.onInitialized(() => _pyrightLanguageProvider.onInitialized());
+
   connection.onRequest(SemanticTokensRequest.type, (params) => {
+    syncIfDocChange(params.textDocument.uri);
     const languageService = getLanguageService(params.textDocument.uri);
 
     return { data: languageService.getTokens() };
   });
 
-  connection.onHover((params) => {
-    const languageService = getLanguageService(params.textDocument.uri);
-
-    return languageService.completionProvider.getHelp(params.position);
+  connection.onHover(async (params, token) => {
+    return await dispatch(params, {
+      async sas(languageService) {
+        return await languageService.completionProvider.getHelp(
+          params.position,
+        );
+      },
+      async python(pyrightLanguageService) {
+        return await pyrightLanguageService.onHover(params, token);
+      },
+    });
   });
 
-  connection.onCompletion((params) => {
-    const languageService = getLanguageService(params.textDocument.uri);
-    completionProvider = languageService.completionProvider;
-
-    return completionProvider.getCompleteItems(params.position);
+  connection.onCompletion(async (params, token) => {
+    return await dispatch(params, {
+      async sas(languageService) {
+        if (params.context?.triggerCharacter !== " ") {
```

Comment:
There's a situation that the `triggerCharacter` is `undefined` (triggered by typing matched keyword), in that case we should allow autocomplete.
```suggestion
        if (params.context?.triggerCharacter && params.context.triggerCharacter !== " ") {
```